### DISCUSSION TITLE
ImportError: cannot import name 'ALREADY_HANDLED' from 'eventlet.wsgi'

### DISCUSSION THREAD

**Discussion by otl1** [miguelgrinberg/Flask-SocketIO#1975] [Q&A]
Hello @miguelgrinberg , thanks for the fantastic library. I have deployed my Flask-SocketIO app on ubuntu and I am using nginx, gunicorn, supervisor. I have installed all the essential latest libraries for my project. In windows my project is working flawlessly but in ubuntu I have been getting `ImportError: cannot import name 'ALREADY_HANDLED' from 'eventlet.wsgi'` 
My `/etc/nginx/sites-enabled/project` configuration is as follow:
```
server {
        listen 80;
        server_name 17.176.170.27;

        include snippets/phpmyadmin.conf;

        location /static {
                alias /home/ubuntu/Test_Server/project/static;
        }

        location / {
                proxy_pass http://localhost:8000;
                include /etc/nginx/proxy_params;
                proxy_redirect off;
        }

        location /socket.io {
                include proxy_params;
                proxy_http_version 1.1;
                proxy_buffering off;
                proxy_set_header Upgrade $http_upgrade;
                proxy_set_header Connection "Upgrade";
                proxy_pass http://localhost:8000/socket.io;
        }

```
My `/etc/supervisor/conf.d/project.conf` configuration is as follow:
```
[program:project]
directory=/home/ubuntu/Test_Server
#command=/home/ubuntu/Test_Server/venv/bin/gunicorn -w 1 run:app
command=/home/ubuntu/Test_Server/venv/bin/gunicorn --worker-class eventlet -w 1 run:app
user=ubuntu
autostart=true
autorestart=true
stopasgroup=true
killasgroup=true
stderr_logfile=/var/log/project/project.err.log
stdout_logfile=/var/log/project/project.out.log
```
My `requirements.txt` as follow:
```
eventlet==0.33.3
Flask==2.2.3
Flask-Bootstrap==3.3.7.1
Flask-MQTT==1.1.1
flask-session2==1.3.1
Flask-SocketIO==5.3.3
Flask-SQLAlchemy==3.0.3
Flask-WTF==1.1.1
greenlet==2.0.2
gunicorn==20.1.0
itsdangerous==2.1.2
Jinja2==3.1.2
MarkupSafe==2.1.2
mysql-connector-python==8.0.33
mysqlclient==2.1.1
paho-mqtt==1.6.1
passlib==1.7.4
protobuf==3.20.3
python-engineio==4.4.1
python-socketio==5.8.0
pytz==2022.7.1
six==1.16.0
SQLAlchemy==2.0.9
typing_extensions==4.5.0
visitor==0.1.3
Werkzeug==2.2.3
WTForms==3.0.1
supervisor==4.2.1
dnspython==2.3.0
bidict==0.22.1
cachelib==0.9.0
click==8.1.3
Jinja2==3.0.3
```
I tried downgrading eventlet but it seems to be also not working. Could you please help me to solve this issue?

**Comment by miguelgrinberg** [comment]
This isn't related to this project. You have incompatible versions of Gunicorn and Eventlet installed. It looks like there is a bug open for this issue already: https://github.com/benoitc/gunicorn/issues/2582

**Reply by Avengm** [reply]
Hi @miguelgrinberg , Thanks for the quick response. I am using latest release of `gunicorn 20.1.0 `and `eventlet 0.33.3 `. Could you please suggest which version of gunicorn and eventlet should I use?

**Reply by miguelgrinberg** [reply]
I'm not sure. You need to downgrade eventlet to a version from around the time of the gunicorn release that you are using.

**Comment by howardai** [comment]
Just want to mention that while I've been searching everywhere, seems like the issue was Gunicorn for my instance and updating it to latest version solved everything. Here's my stack of working version modules:
dnspython==2.4.2
eventlet==0.34.3
Flask==2.0.3
Flask-SocketIO==4.3.1
gunicorn==21.2.0
